### PR TITLE
Multicore systemtests

### LINKED ISSUE #23285: SystemTests is slow, they should be run on multiple cores
### Expected behavior

`systemtests` should be run on multiple cores, allocating a small number of threads to each running test.
Initial tests show that using 12 cores, runtime goes down from ~2h to ~18min.

### Actual behavior

`systemtests` is run serially, and each test has access to all the machine's threads.

### Steps to reproduce the behavior

Run `./systemtests` in the `build` directory.

### Platforms affected

All

### PR DESCRIPTION
**Pull request by nvaytet**
**Description of work.**

Using the `multiprocessing` module to launch the `TestManager` with `Process`.
Runtimes for `./systemtests` has gone down from ~2h to ~18min using 12 cores.

**To test:**

In the `build` directory, `./systemtests -j 12`

Fixes #23285.
---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by nvaytet** [comment]
Hi @martyngigg I think i've gotten to a stage where running the `systemtests` in parallel seems to work in a stable way.
I had problems where I was distributing the tests among the CPUs, that some where deleting data that other CPUs still needed. So now I am giving out the tests in batches (all the tests in the same `.py` file get sent to one CPU) and this has fixed the issue.
I have also created a second output mode (`-q` or `--quiet`) which prints only the progress to terminal rather than the full log (the full log is still printed by default). This looks something like this:
```
>$ time ./systemtest -l "notice" -j 12 --showskipped --quiet
Updating testing data...
[100%] Built target StandardTestData
[100%] Built target SystemTestData

Running tests...
FrameworkManager-[Notice] Welcome to Mantid 3.13.20180820.2132
FrameworkManager-[Notice] Please cite: http://dx.doi.org/10.1016/j.nima.2014.07.029 and this release: http://dx.doi.org/10.5286/Software/Mantid
========================================
Number of tests run per CPU:
CPU 0 will execute 37 tests
CPU 1 will execute 37 tests
CPU 2 will execute 37 tests
CPU 3 will execute 36 tests
CPU 4 will execute 36 tests
CPU 5 will execute 36 tests
CPU 6 will execute 36 tests
CPU 7 will execute 36 tests
CPU 8 will execute 36 tests
CPU 9 will execute 36 tests
CPU 10 will execute 36 tests
CPU 11 will execute 36 tests
========================================
[  0%]   1/435 : DOSTest.DOSCastepTest .................................................................................. (success: 0.05s)
[  0%]   2/435 : ISISIndirectBayesTest.JumpCETest ....................................................................... (success: 0.06s)
[  0%]   3/435 : ISISIndirectInelastic.IRISCalibration .................................................................. (success: 0.03s)
[  0%]   4/435 : HFIRTransAPIv2.HFIRTrans1 .............................................................................. (success: 1.30s)
[  1%]   5/435 : DOSTest.DOSIRActiveTest ................................................................................ (success: 0.04s)
[  1%]   6/435 : ISISIndirectBayesTest.JumpFickTest ..................................................................... (success: 0.06s)
[  1%]   7/435 : AbinsTest.AbinsBinWidth ................................................................................ (success: 1.65s)
[  1%]   8/435 : ISIS_PowderPearlTest.CreateCalTest ..................................................................... (success: 1.65s)
[  2%]   9/435 : ISISIndirectInelastic.IRISConvFit ...................................................................... (success: 0.56s)
[  2%]  10/435 : LiquidsReflectometryReductionWithBackgroundTest.BadDataTOFRangeTest .................................... (success: 2.94s)
[  2%]  11/435 : DOSTest.DOSPartialCrossSectionScaleTest ................................................................ (success: 0.23s)
[  2%]  12/435 : ISISIndirectBayesTest.JumpHallRossTest ................................................................. (success: 0.07s)
[  2%]  13/435 : ISISIndirectInelastic.IRISDiagnostics .................................................................. (success: 0.03s)
[  3%]  14/435 : HFIRTransAPIv2.HFIRTrans2 .............................................................................. (success: 0.83s)
[  3%]  15/435 : DOSTest.DOSPartialSummedContributionsCrossSectionScaleTest ............................................. (success: 0.15s)
[  3%]  16/435 : ISISIndirectBayesTest.JumpTeixeiraTest ................................................................. (success: 0.07s)
[  3%]  17/435 : ISISIndirectInelastic.IRISElwinAndMSDFit ............................................................... (success: 0.29s)
[  4%]  18/435 : MagnetismReflectometryReductionTest.MRFilterCrossSectionsTest .......................................... (success: 5.30s)
[  4%]  19/435 : DOSTest.DOSPartialSummedContributionsTest .............................................................. (success: 0.16s)
[  4%]  20/435 : ISISIndirectBayesTest.QLDataTest ....................................................................... (skipped:  -- s)
[  4%]  21/435 : VesuvioCorrectionsTest.TestCorrectionsInBackScatteringSpectra .......................................... (success: 7.63s)
[  5%]  22/435 : HFIRTransAPIv2.HFIRTransmissionBeamCenter .............................................................. (success: 1.26s)
[  5%]  23/435 : ISISIndirectBayesTest.QLResNormTest .................................................................... (skipped:  -- s)
[  5%]  24/435 : DOSTest.DOSPartialTest ................................................................................. (success: 0.16s)
.
.
.
```

All I am missing now is finding the write place to add the `-j` option (probably somewhere in the `buildconfig` folder) so that the `systemtests` are running in parallel on the build servers. Would you be able to point me to the right place? Thanks!

**Comment by nvaytet** [comment]
Ok I am still having problems with tests failing because the same data is used over multiple python files. I am looking for a way to run tests in a separate directory for each core. I have reverted to a "In Progress" state.

**Comment by martyngigg** [comment]
@nvaytet This looks good so far. I think we'll also need to update [SystemTests/scripts/InstallerTests.py](
https://github.com/mantidproject/mantid/blob/1c1c4437800ba3e32ae16dc073dfe94316c1739c/Testing/SystemTests/scripts/InstallerTests.py) with the `-j` option and feed it in from [buildconfig/Jenkins/systemtests](https://github.com/mantidproject/mantid/blob/1c1c4437800ba3e32ae16dc073dfe94316c1739c/buildconfig/Jenkins/systemtests) & [buildconfig/Jenkins/systemtests.bat](https://github.com/mantidproject/mantid/blob/1c1c4437800ba3e32ae16dc073dfe94316c1739c/buildconfig/Jenkins/systemtests.bat) as these are the scripts Jenkins uses as it runs the tests on a packaged copy.

I think I might add a maintenance task to try and amalgamate some of this stuff.

File: Testing/SystemTests/lib/systemtests/stresstesting.py [hunk 1]
```diff
@@ -708,42 +742,123 @@ class TestManager(object):
     '''
 
     def __init__(self, test_loc, runner, output = [TextResultReporter()],
-                 testsInclude=None, testsExclude=None, exclude_in_pr_builds=None):
+                 quiet=False, testsInclude=None, testsExclude=None,
+                 exclude_in_pr_builds=None, showSkipped=False,
+                 output_on_failure=False, clean=False,
+                 test_count=[0,0,0], process_number=0, ncores=1):
```
Comment (martyngigg) [inline Testing/SystemTests/lib/systemtests/stresstesting.py]: Can we avoid lists as default arguments as it's generally considered bad practice: https://docs.python-guide.org/writing/gotchas/#what-you-wrote

File: Testing/SystemTests/scripts/InstallerTests.py [hunk 2]
```diff
@@ -76,8 +78,9 @@
 log("Running system tests. Log files are: '%s' and '%s'" % (testRunLogPath,testRunErrPath))
 try:
     run_test_cmd = '%s %s %s/runSystemTests.py --loglevel=%s --executable="%s" --exec-args="%s"' % \
-                (installer.python_cmd,  installer.python_args,
-                 THIS_MODULE_DIR, options.log_level, installer.python_cmd, installer.python_args)
+                (installer.python_cmd, installer.python_args, THIS_MODULE_DIR,
+                options.log_level, installer.python_cmd, installer.python_args)
+    run_test_cmd += " -j=%i --quiet --output-on-failure" % options.ncores
```
Comment (martyngigg) [inline Testing/SystemTests/scripts/InstallerTests.py]: The `-j=` part here doesn't seem to work as we would like. The PR test log shows: 

```
python /home/builder/jenkins-linode/workspace/pull_requests-rhel7/Testing/SystemTests/scripts/InstallerTests.py -o -d /home/builder/jenkins-linode/workspace/pull_requests-rhel7/build -j 13 --exclude-in-pull-requests
Usage: runSystemTests.py [options]
 
runSystemTests.py: error: option -j: invalid integer value: '=13'
```

**Review by martyngigg — CHANGES_REQUESTED** [review]
Thanks for this. The code changes look good to me.

I've just one minor comment plus I think a tweak that is needed to `InstallerTests.py` to get it to use `runSystemTests.py` correctly.

**Comment by nvaytet** [comment]
Hi Martyn, thanks for the comments. I decided to run the systemtests with the new `--quiet` output mode, but also with the `--output-on-failure` flag to print the full log in case the test result is not `success`. This should minimize the size of the logs but still have all the info in case a test has failed. Do you think this is the right thing to do or do you prefer to keep the full log?

**Comment by martyngigg** [comment]
Hi Neil. I think that make sense. It's how we run the unit tests with `ctest` and I can't really think why we want the output if the tests passed.

**Comment by martyngigg** [comment]
The current system test failure on `master` provides a nice way to check the behaviour of the failure case.

**Comment by nvaytet** [comment]
Ah yes, perfect testing platform! The XML reporter seems to be working as we want it to, picking up the results in the different `TEST-systemtests-*.xml` files

**Review by martyngigg — CHANGES_REQUESTED** [review]
Could we update the [developer documentation](/mantidproject/mantid/blob/master/dev-docs/source/SystemTests.rst) to describe the new options?

**Comment by martyngigg** [comment]
I was just playing around with the `systemtest` script and I tried: `./systemtest -R Focus -j8 --quiet`. It seems to execute the expected tests but also prints several lines containing

```No tests defined in /media/data1/source/github/mantidproject/mantid/Testing/SystemTests/tests/analysis. Please ensure all test classes sub class stresstesting.MantidStressTest.```

that I think could cause confusion. Do you know where they might come from?

Full output:
```
$ ./systemtest -R Focus -j8 --quiet
Updating testing data...
ninja: no work to do.
ninja: no work to do.

Running tests...
ConfigService-[Information] Unable to locate directory at: /etc/mantid/instrument
ConfigService-[Information] This is Mantid version 3.13.20180829.844 revision g67bdf45
ConfigService-[Information] running on ndw1213 starting 2018-08-29T10:32Z
ConfigService-[Information] Properties file(s) loaded: /media/data1/source/github/mantidproject/mantid-builds/nightly-py2/bin/Mantid.properties, /etc/mantid.local.properties, /home/dmn58364/.mantid/Mantid.user.properties
ConfigService-[Information] Unable to locate directory at: /etc/mantid/instrument
FrameworkManager-[Notice] Welcome to Mantid 3.13.20180829.844
FrameworkManager-[Notice] Please cite: http://dx.doi.org/10.1016/j.nima.2014.07.029 and this release: http://dx.doi.org/10.5286/Software/Mantid
FrameworkManager-[Information] Instrument updates disabled - cannot update instrument definitions.
FrameworkManager-[Information] Version check disabled.
No tests defined in /media/data1/source/github/mantidproject/mantid/Testing/SystemTests/tests/analysis. Please ensure all test classes sub class stresstesting.MantidStressTest.
No tests defined in /media/data1/source/github/mantidproject/mantid/Testing/SystemTests/tests/analysis. Please ensure all test classes sub class stresstesting.MantidStressTest.
No tests defined in /media/data1/source/github/mantidproject/mantid/Testing/SystemTests/tests/analysis. Please ensure all test classes sub class stresstesting.MantidStressTest.
No tests defined in /media/data1/source/github/mantidproject/mantid/Testing/SystemTests/tests/analysis. Please ensure all test classes sub class stresstesting.MantidStressTest.
No tests defined in /media/data1/source/github/mantidproject/mantid/Testing/SystemTests/tests/analysis. Please ensure all test classes sub class stresstesting.MantidStressTest.
No tests defined in /media/data1/source/github/mantidproject/mantid/Testing/SystemTests/tests/analysis. Please ensure all test classes sub class stresstesting.MantidStressTest.
```

**Comment by martyngigg** [comment]
@nvaytet Thanks for the explanation - that makes sense. We had similar troubles with the unit tests and spent time weeding out all of the tests that wrote to files with the same name. Perhaps this is something we could address in maintenance to improve things further?

I think the behaviour you have at the moment is fine but if we could avoid printing the warning then I think would help developers who might become confused.

**Comment by nvaytet** [comment]
Ok I will update the documentation now.

File: dev-docs/source/SystemTests.rst [hunk 3]
```diff
@@ -212,6 +212,89 @@ e.g.
 
 would run all of the tests whose name contains SNS.
 
+Running the tests on multiple cores
```
Comment (martyngigg) [inline dev-docs/source/SystemTests.rst]: :+1:

**Review by martyngigg — APPROVED** [review]
This looks really good to me.

On one of the 8 core nodes a whole PR job timing when from ~2hrs to 50mins. A great improvement! :beers:

**Comment by nvaytet** [comment]
I would still like to get the optimal speedup by making sure all the tests are writing to a single directory whose path can be customized for each core.
As for the failing System Test, I guess I'll just wait for a fix and then merge from `master`.

**Comment by nvaytet** [comment]
I created a maintenance task for the directories in which the tests write: #23373

**Review by rosswhitfield — CHANGES_REQUESTED** [review]

File: Testing/SystemTests/lib/systemtests/stresstesting.py [hunk 4]
```diff
@@ -708,42 +742,123 @@ class TestManager(object):
...
+
+        # Now we distribute the tests to each core
+        # This is done by sorting the modules by descending order of number of tests
+        # We then iterate through that list and give all the tests inside a given module
+        # to the core which currently has the lowest number of tests.
+        # The number of tests for that core is then incremented by the number of tests
+        # inside the module it has just received.
+        ntests_per_core = [0] * ncores
+        self._tests = []
+        for key, value in sorted(modcounts.iteritems(), key=lambda (k,v): (v,k),reverse=True):
```
[the remark this answers is not in the document]
Comment (rosswhitfield) [reply Testing/SystemTests/lib/systemtests/stresstesting.py]: This fixes it
```diff
 import tempfile
 import time
 import unittest
-from six import PY3
+from six import PY3, iteritems
 
 # Path to this file
 THIS_MODULE_DIR = os.path.dirname(os.path.realpath(__file__))
@@ -832,7 +832,7 @@ class TestManager(object):
         # inside the module it has just received.
         ntests_per_core = [0] * ncores
         self._tests = []
-        for key, value in sorted(modcounts.iteritems(), key=lambda (k,v): (v,k),reverse=True):
+        for key, value in sorted(iteritems(modcounts), key=lambda k_v: (k_v[1],k_v[0]), reverse=True):
             for i in range(ncores):
                 if(ntests_per_core[i] == min(ntests_per_core)):
                     ntests_per_core[i] += value
```

[the remark this answers is not in the document]
Comment (nvaytet) [reply Testing/SystemTests/lib/systemtests/stresstesting.py]: Hi, thanks for this! Does the rest then run ok with Python3?

[the remark this answers is not in the document]
Comment (rosswhitfield) [reply Testing/SystemTests/lib/systemtests/stresstesting.py]: I think so. I didn't thoroughly test it, I just made sure it runs.

**Review by rosswhitfield — APPROVED** [review]

**Pull request merged**